SQL: Redact credentials in connection exceptions

bpintea opened the pull request:
This PR adds the functionality to redact the credentials from the
exceptions generated when a connection attempt fails, preventing them
from leaking into logs, console history etc.

There are a few causes that can lead to failed connections. The most
challenging to deal with is a malformed connection string. The redaction
tries to get around it by modifying the URI to a parsable state, so that
the redaction can be applied reliably. If there's no reliability
guarantee, the redaction will bluntly replace the entire connection
string and the user informed about the option to modify it so that the
redaction won't apply. (This is done by using a caplitalized scheme,
which is legal, but otherwise never used in practice.)

The PR fixes a couple of other issues with the URI parser:
- it allows an empty hostname, or even entire connection string (as per
the existing documentation);
- it reduces the editing of the connection string in the exception
messages (so that the user easier recognize their input);
- it uses the default URI as source for the scheme and hostname.

Closes #56474

bpintea committed:
Redact credentials in connection exceptions

This commit adds the functionality to redact the credentials from the
exceptions generated when a connection attempt fails, preventing them
from leaking into logs, console history etc.

There are a few causes that can lead to failed connections. The most
challenging to deal with is a malformed connection string. The redaction
tries to get around it by modifying the URI to a parsable state, so that
the redaction can be applied reliably. If there's no reliability
guarantee, the redaction will bluntly replace the entire connection
string and the user informed about the option to modify it so that the
redaction won't apply. (This is done by using a caplitalized scheme,
which is legal, but otherwise never used in practice.)

The commit fixes a couple of other issues with the URI parser:
- it allows an empty hostname, or even entire connection string (as per
the existing documentation);
- it reduces the editing of the connection string in the exception
messages (so that the user easier recognize their input);
- it uses the default URI as source for the scheme and hostname.

bpintea commented:
ok to test

costin reviewed (approved):
Looks good. One suggestion I have is instead of using `*` for each character (which leaks information on how long the redacted data is) I would just use something like `<redacted>` or `[hidden]` or `<removed>` - essentially something that stands out while preserving the URI correctness to some degree.
This way regardless of how long the password/user is, it should be clear the URI sent.

bpintea commented:
> instead of using * for each character [...] use something like `<redacted>` or `[hidden]` or `<removed>`

I consider this initially too, there's a debatable downside to it, though: the `URISynthaxException` hints where's the problem with a character offset. For instance:  `http://user@host/pa^th` will generate `Illegal character in path at index 19: http://user@host/pa^th`. The index can of course be also updated, but that'd modify the string to a degree that it might be confusing for the user if this index is relevant for the original string they provided, or the updated one after redaction. And since we can't provide visual hints (of the type `-------^`) without modifying the string even more, I defaulted to the `*`-replacement, which seems generally as an acceptable solution.

NB. This provided index isn't always hinting to the exact character, it can also generally indicate the affected part of the URI: `http://user@[1:1:1]/path` -> `IPv6 address too short at index 13: http://user@[1:1:1]/path` (i.e. `[`).

matriv commented on x-pack/plugin/sql/sql-client/src/test/java/org/elasticsearch/xpack/sql/client/UriUtilsTests.java:
```diff
@@ -59,15 +82,197 @@ public void testHttpQuery() throws Exception {
+
+    public void testNoRedaction() {
+        assertEquals(
+                "Invalid connection configuration: Illegal character in fragment at index 16: HTTP://host#frag#ment",
```
Minor: Those 2 lines seem to have inconsistent identation compared to the rest of the file.

bpintea replied:
Ah, rite. Thanks!
This file had a bit of a styles mix. I've aligned it to what seems to have been an initial one, but happy to re-adjust it if I got it wrong.

matriv reviewed (approved):
LGTM. Nice! Left a minor formatting comment.

bpintea committed:
style corrections

Allign code in file w.r.t.:
- indentation size;
- parameters placement;
- closing paranthesis placement.

bpintea commented:
@elasticmachine run elasticsearch-ci/1 elasticsearch-ci/2 elasticsearch-ci/bwc

bpintea commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into enh/redact_credentials_in_invalid_uri

astefan commented on x-pack/plugin/sql/sql-client/src/main/java/org/elasticsearch/xpack/sql/client/UriUtils.java:
```diff
@@ -42,42 +56,211 @@ public static URI parseURI(String connectionString, URI defaultURI) {
+                    return editURI(uri, faults, hasPort);
+                } catch (URISyntaxException use) {
+                    int idx = use.getIndex();
+                    if (idx < 0 || idx >= sb.length()) {
```
When can the `idx` be `> sb.length()`? And why the equality is not a valid use case to be redacted?

bpintea replied:
> When can the `idx` be > `sb.length()`?

_Probably_ never, but not sure what URISyntaxException can provide as "valid" value there.

> And why the equality is not a valid use case to be redacted?

In this case the index would fall outside the input string length; for instance parsing "http://" as URI will return `Expected authority at index 7: http://`, for which the equality holds.

astefan reviewed (approved):
LGTM. Nice work!
I've left some minor comments/questions.

bpintea committed:
Address review comments

- remove unused var;
- complete code comment.

bpintea commented on x-pack/plugin/sql/jdbc/src/main/java/org/elasticsearch/xpack/sql/jdbc/JdbcConfiguration.java:
```diff
@@ -115,15 +115,15 @@ public static JdbcConfiguration create(String u, Properties props, int loginTime
 
     private static URI parseUrl(String u) throws JdbcSQLException {
         String url = u;
```
good point, thanks.

bpintea commented on x-pack/plugin/sql/sql-client/src/main/java/org/elasticsearch/xpack/sql/client/UriUtils.java:
```diff
@@ -7,18 +7,32 @@
+    static final String HTTPS_PREFIX = HTTPS_SCHEME + "://";
+
     /**
      * Parses the URL provided by the user and
```
Sure, makes sense.

bpintea commented on x-pack/plugin/sql/sql-client/src/main/java/org/elasticsearch/xpack/sql/client/UriUtils.java:
```diff
@@ -42,42 +56,211 @@ public static URI parseURI(String connectionString, URI defaultURI) {
+        }
+
+        public static String redactCredentialsInConnectionString(String connectionString) {
+            if (connectionString.startsWith(HTTP_PREFIX.toUpperCase(Locale.ROOT))
```
> Do we have a good/strong argument in favor of "deactivating" the redaction if URLs when the schema is uppercase and activating it when it's lowercase?

Just a small clarification: redaction is currently active in all cases, except an all-uppercase schema.

As @astefan pointed out, this can be an issue if the user provides it like that unknowingly initially (potentially generated all-uppercase by software). Hopefully this happens very rarely, but we can choose something else if we are concerned about it (like some `HttP` camelcase or even a completely different schema, though that might be a bit confusing).

> Wouldn't a URL parameter do the job, similar to `validate.properties`?

The trick here would be to reliably recognise that parameter in case the URI itself is broken. We could also do a simple string scanning (`lower_case_uri.contains("redaction=false")`) and hope for no false positives - should be indeed very unlikely -, but the downside of that is that the user would potentially no longer get the same error message if offsets change (which is however a possibility with current mechanism as well, in case the user didn't provide a schema in the first place), or they could introduce new errors.

astefan replied:
We would default the `redaction` parameter to `true` and if the the URL is broken and the parameter is not recognizable the URL sensitive data wouldn't leak anyway. If the user is interested in disabling the redaction step, then will look into the URL correctness anyway.

bpintea replied:
Yep, that could also work.
Myself I'm still biased towards the existing implementation as it's slightly simpler and doesn't require an extra param, but "close to the fence", so if anyone would have a swaying opinion I'll change to the above (or potentially new) suggestion. CC: @matriv @costin

matriv replied:
Personally I'd vote to keep it as is instead of introducing a new parameter. Do we have any experience/demand in the past for this uppercase URIs? To me seems like a super edge case and wouldn't invest too much to have the perfect handling for it.

bpintea commented:
@elasticmachine run elasticsearch-ci/bwc

bpintea commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into enh/redact_credentials_in_invalid_uri

bpintea commented:
@elasticmachine run elasticsearch-ci/2

bpintea commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into enh/redact_credentials_in_invalid_uri
